gammapkt.cc

```diff
@@ -1046,13 +1048,93 @@ void wollaeger_thermalisation(Packet &pkt) {
   }
 }
 
+void guttman_thermalisation(Packet &pkt) {
+  // Guttman+2024, arXiv:2403.08769v1
+  // extension of the Wollaeger scheme. Rather than calculating a single optical depth in radial outward
+  // direction, it calculates a spherical average in all possible gamma-ray emission directions.
+
+  // toy value for the mean gamma ray opacity for a start, the paper discuss this some more detail, too
+  constexpr double mean_gamma_opac =
+      0.03;  // mean gamma opacity from section 3.2, taken the lower one due to almost symmetrical matter at late times
+
+  const double t_0 = globals::tmin;
+  const double t = pkt.prop_time;
+
+  // calculate average optical w.r.t. to emission direction
+  // discretize the two sphere into octants, i.e. four values in phi and two in theta
+  const int numb_rnd_dirs = 100;
+  // calculate all column densities and their average
+  double column_densities[numb_rnd_dirs] = {};
+  double avg_column_density = 0.;
+  for (int i = 0; i < numb_rnd_dirs; i++) {
+    // compute column density by moving an artificial packet outwards and integrating over local density
+    // WARNING: This simple implementation relies on a relatively large number of random directions
+    // as it assumes that the directions are distributed equally which is likely not the case with low statistics.
+
+    // step 1: draw a random direction
+    Packet pkt_copy = pkt;
+    // phi rotation: around z-axis
+    std::array<double, 3> random_dir = get_rand_isotropic_unitvec();
+    pkt_copy.dir = random_dir;  // fix new direction
+
+    // step 2: move packet into the calculated direction and integrate the density
+    bool end_packet = false;
+    while (!end_packet) {
+      // distance to the next cell
+      const auto [sdist, snext] =
+          grid::boundary_distance(pkt_copy.dir, pkt_copy.pos, pkt_copy.prop_time, pkt_copy.where, &pkt_copy.last_cross);
+      const double s_cont = sdist * std::pow(t, 3.) / std::pow(pkt_copy.prop_time, 3.);
+      const int mgi = grid::get_cell_modelgridindex(pkt_copy.where);
+      if (mgi != grid::get_npts_model()) {
+        column_densities[i] += grid::get_rho_tmin(mgi) * s_cont;  // contribution to the integral
+      }
+      // move packet copy now
+      move_pkt_withtime(pkt_copy, sdist);
+
+      grid::change_cell(pkt_copy, snext);
+      end_packet = (pkt_copy.type == TYPE_ESCAPE);
+    }
+    avg_column_density += column_densities[i];
+  }
+  avg_column_density /= numb_rnd_dirs;  // calculate average here
```

Comment:
A minor suggestion: With column_densities as a std::array, you can avoid having a mutable state of avg_columns_density:
`const double avg_column_density = std::accumulate(column_densities.cbegin(), column_densities.cend()) / std::ssize(column_densities);`